Fix various issues in DDLWorker (SIGSEGV and others)

azat opened the pull request:
Changelog category (leave one):
- Not for changelog

Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Fix various issues in DDLWorker (SIGSEGV and others)

Introduced-in: #16193
Cc: @tavplubix

azat committed:
DDLWorker: Fix reference to zookeeper with distributed_ddl.pool_size > 1 (thread pool)

azat committed:
DDLTask: Use std::uncaught_exceptions() over deprecated std::uncaught_exception()

azat committed:
tests: regression for test for invalid reference to zookeeper for distributed_ddl.pool_size > 1

azat committed:
tests: catch exceptions in threads in test_distributed_ddl_parallel

Otherwise the test will not report an error.

azat committed:
tests: decrease distributed_ddl_task_timeout to 60 in test_distributed_ddl_parallel

azat committed:
DDLWorker: avoid comparing pointers for getting min element

The current_tasks always updated with emplace_back, so front() is ok.

azat committed:
DDLWorker: remove only completed tasks (significant for distributed_ddl.pool_size > 1)

Otherwise it will SIGSEGV due to invalid-read (UAF).

azat commented:
By the way the reason this wasn't captured by tests (but you can see that the test was executed too long, and [logs](https://clickhouse-test-reports.s3.yandex.net/16193/863c0992540c68b781b393a35d8c8f47dddbdd20/integration_tests_(thread)/integration_run_test_distributed_ddl_parallel_test_py_0.txt.out.log) contains exceptions) is that the query was executed in the threads and those exceptions was simply ignored... This patch set will address this.

tavplubix commented on src/Interpreters/DDLWorker.cpp:
```diff
@@ -338,10 +338,10 @@ void DDLWorker::scheduleTasks()
         String min_entry_name = last_skipped_entry_name ? std::min(min_task->entry_name, *last_skipped_entry_name) : min_task->entry_name;
         begin_node = std::upper_bound(queue_nodes.begin(), queue_nodes.end(), min_entry_name);
-        current_tasks.clear();
+        current_tasks.remove_if([](const DDLTaskPtr & t) { return t->completely_processed.load(); });
```
Why should we save tasks which was not executed? If task was executed, but its status was not written to ZooKeeper (due to connection loss), then we call `processTask(...)` [here](https://github.com/ClickHouse/ClickHouse/blob/6946514c0b3165ab79de7a9feca5d19d0a73c03e/src/Interpreters/DDLWorker.cpp#L321) to write the status. If task was not executed, then we will read the task from ZooKeeper again and retry it.

azat replied:
Reference to current_tasks is used in processTasks, in case of distirbuted_ddl.pool_size > 1 they are processed in background and such items cannot be removed from the current_tasks.

But this does not solve iterating over those tasks again, seems that some pending state should be introduced for this async execution, what do you think?

tavplubix replied:
> Reference to current_tasks is used in processTasks

You're right, I have removed `reset_state()` from [here](https://github.com/ClickHouse/ClickHouse/blob/6946514c0b3165ab79de7a9feca5d19d0a73c03e/src/Interpreters/DDLWorker.cpp#L1009), but forgot to add something like `if (worker_pool) worker_pool->wait()` for the case when `pool_size > 1`.

> But this does not solve iterating over those tasks again, seems that some pending state should be introduced for this async execution, what do you think?

I think that we should not have any background tasks running if main thread was restarted/reinitialied due to connection loss or some unexpected exception.

azat replied:
>You're right, I have removed reset_state() from here, but forgot to add something like if (worker_pool) worker_pool->wait() for the case when pool_size > 1.

Yes, this also should be addressed.

>I think that we should not have any background tasks running if main thread was restarted/reinitialied due to connection loss or some unexpected exception.

If the main thread was restarted then yes, but that code called every time not only after restarting the main thread, am I missing something? (and also the code above that iterates over current_tasks too)

tavplubix commented on src/Interpreters/DDLWorker.cpp:
```diff
@@ -338,10 +338,10 @@ void DDLWorker::scheduleTasks()
     {
         /// We will recheck status of last executed tasks. It's useful if main thread was just restarted.
-        auto & min_task = *std::min_element(current_tasks.begin(), current_tasks.end());
+        auto & min_task = current_tasks.front();
```
Ok, but let's add an assertion that the first task is the minimum one

azat replied:
By entry name over all not completed but started tasks?
Looks complex (tasks to current_tasks always added to the end, so this should not be a problem)

tavplubix replied:
Yes, I mean something like
`assert(min_task == *std::min_element(current_tasks.begin(), current_tasks.end()));`

azat replied:
But this is not correct, since they will be compared by pointer address, and there is no strict guarantee that the newly allocated pointer will be greater the previous

azat committed:
DDLWorker: wait for pending async tasks

azat committed:
DDLWorker: optimize processing of unfinished tasks

Do not look at zookeeper if was_executed is false

azat committed:
DDLWorker: process unfinished tasks only after reinitialization

azat committed:
Load dictionaries before starting accepting connections and DDLWorker

azat committed:
DDLWorker: avoid NULL dereference on termination and failed zookeeper initialization

Log snipped shows the problem:

    2021.02.24 04:40:29.349181 [ 39 ] {} <Warning> DDLWorker: DDLWorker is configured to use multiple threads. It's not recommended because queries can be reordered. Also it may cause some unknown issues to appear.
    2021.02.24 04:40:29.349516 [ 39 ] {} <Information> Application: Ready for connections.
    2021.02.24 04:40:29.349602 [ 74 ] {} <Debug> DDLWorker: Started DDLWorker cleanup thread
    2021.02.24 04:40:29.349639 [ 73 ] {} <Debug> DDLWorker: Starting DDLWorker thread
    2021.02.24 04:40:29.349698 [ 73 ] {} <Debug> DDLWorker: Started DDLWorker thread
    2021.02.24 04:40:29.352548 [ 73 ] {} <Error> virtual void DB::DDLWorker::initializeMainThread(): Code: 999, e.displayText() = Coordination::Exception: All connection tries failed while connecting to ZooKeeper. nodes: 192.168.112.3:2181
    Poco::Exception. Code: 1000, e.code() = 111, e.displayText() = Connection refused (version 21.3.1.1), 192.168.112.3:2181
    Poco::Exception. Code: 1000, e.code() = 111, e.displayText() = Connection refused (version 21.3.1.1), 192.168.112.3:2181
    Poco::Exception. Code: 1000, e.code() = 111, e.displayText() = Connection refused (version 21.3.1.1), 192.168.112.3:2181
     (Connection loss), Stack trace (when copying this message, always include the lines below):

    0. Coordination::Exception::Exception(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, Coordination::Error, int) @ 0xfe93923 in /usr/bin/clickhouse
    1. Coordination::Exception::Exception(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, Coordination::Error) @ 0xfe93ba2 in /usr/bin/clickhouse
    2. Coordination::ZooKeeper::connect(std::__1::vector<Coordination::ZooKeeper::Node, std::__1::allocator<Coordination::ZooKeeper::Node> > const&, Poco::Timespan) @ 0xfed3a01 in /usr/bin/clickhouse
    3. Coordination::ZooKeeper::ZooKeeper(std::__1::vector<Coordination::ZooKeeper::Node, std::__1::allocator<Coordination::ZooKeeper::Node> > const&, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, Poco::Timespan, Poco::Timespan, Poco::Timespan) @ 0xfed2222 in /usr/bin/clickhouse
    4. zkutil::ZooKeeper::init(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, int, int, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&) @ 0xfe961cd in /usr/bin/clickhouse
    5. zkutil::ZooKeeper::ZooKeeper(Poco::Util::AbstractConfiguration const&, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&) @ 0xfe97a96 in /usr/bin/clickhouse
    6. void std::__1::allocator_traits<std::__1::allocator<zkutil::ZooKeeper> >::__construct<zkutil::ZooKeeper, Poco::Util::AbstractConfiguration const&, char const (&) [10]>(std::__1::integral_constant<bool, true>, std::__1::allocator<zkutil::ZooKeeper>&, zkutil::ZooKeeper*, Poco::Util::AbstractConfiguration const&, char const (&) [10]) @ 0xed98387 in /usr/bin/clickhouse
    7. DB::Context::getZooKeeper() const @ 0xed75190 in /usr/bin/clickhouse
    8. DB::DDLWorker::getAndSetZooKeeper() @ 0xedb81c9 in /usr/bin/clickhouse
    9. DB::DDLWorker::initializeMainThread() @ 0xedc9eb0 in /usr/bin/clickhouse
    10. DB::DDLWorker::runMainThread() @ 0xedb5d01 in /usr/bin/clickhouse
    11. ThreadFromGlobalPool::ThreadFromGlobalPool<void (DB::DDLWorker::*)(), DB::DDLWorker*>(void (DB::DDLWorker::*&&)(), DB::DDLWorker*&&)::'lambda'()::operator()() @ 0xedcafa1 in /usr/bin/clickhouse
    12. ThreadPoolImpl<std::__1::thread>::worker(std::__1::__list_iterator<std::__1::thread, void*>) @ 0x892651f in /usr/bin/clickhouse
    13. ? @ 0x8929fb3 in /usr/bin/clickhouse
    14. start_thread @ 0x8ea7 in /lib/x86_64-linux-gnu/libpthread-2.31.so
    15. __clone @ 0xfddef in /lib/x86_64-linux-gnu/libc-2.31.so
     (version 21.3.1.1)
    ...
    2021.02.24 04:40:30.025278 [ 41 ] {} <Trace> BaseDaemon: Received signal 15
    2021.02.24 04:40:30.025336 [ 41 ] {} <Information> Application: Received termination signal (Terminated)
    ...
    2021.02.24 04:40:30.582078 [ 39 ] {} <Information> Application: Closed all listening sockets.
    2021.02.24 04:40:30.582124 [ 39 ] {} <Information> Application: Closed connections.
    2021.02.24 04:40:30.583770 [ 39 ] {} <Information> Application: Shutting down storages.
    2021.02.24 04:40:30.583932 [ 39 ] {} <Information> Context: Shutdown disk data
    2021.02.24 04:40:30.583951 [ 39 ] {} <Information> Context: Shutdown disk default
    2021.02.24 04:40:30.584163 [ 46 ] {} <Trace> SystemLog (system.query_log): Terminating
    2021.02.24 04:40:30.586025 [ 39 ] {} <Trace> BackgroundSchedulePool/BgSchPool: Waiting for threads to finish.
    2021.02.24 04:40:34.352701 [ 73 ] {} <Debug> DDLWorker: Initialized DDLWorker thread
    2021.02.24 04:40:34.352758 [ 73 ] {} <Debug> DDLWorker: Scheduling tasks

azat committed:
Add other distributed_ddl settings into config.xml as an example

azat committed:
DDLWorker: Fix processing of tasks in parallel (correct queue begin)

Otherwise it will:
- start from incorrect task and bail because finished node already
  exists
- and also process alreayd processed items

azat committed:
Improve test_distributed_ddl_parallel to cover more cases

Refs: #21264

azat commented:
>Integration tests flaky check (asan) — Flaky tests found: fail: 1, passed: 5, error: 1

Should be addressed in #21264

azat commented:
>Should be addressed in #21264 (after nightly stress testing I cannot reproduce the failure anymore after applying patch from the referenced PR)

Rebased against upstream (that includes that PR)

azat commented:
>Stress test (thread) — Killed by signal (in clickhouse-server.log)

#21305 (`ZstdDeflatingWriteBuffer`)

>Integration tests (release) — fail: 0, passed: 460, error: 26

```
E           ERROR: for hdfs1  Cannot start service hdfs1: driver failed programming external connectivity on endpoint rootteststoragehdfs_hdfs1_1 (c19f177bf04279dc920c4d5cfea6da1ea124a104e1f0f6eff282c16176df690e): Error starting userland proxy: listen tcp4 0.0.0.0:50070: bind: address already in use
```

Should be addressed by #20393
Does not looks related anyway

tavplubix commented on tests/integration/test_distributed_ddl_parallel/test.py:
```diff
@@ -49,17 +73,32 @@ def inner(*args, **kwargs):
+    })
+
+def check_log():
+    # ensure that none of tasks processed multiple times
```
FYI https://github.com/ClickHouse/ClickHouse/blob/184ec67dac727f89702ce12db5d7b51a8dfc2f25/tests/integration/test_distributed_ddl/cluster.py#L104

tavplubix committed:
Update DatabaseReplicatedWorker.cpp

tavplubix commented:
Integration tests (thread) - `test_storage_kafka/test.py::test_kafka_virtual_columns2` - flaky

alexey-milovidov commented:
@azat The issues were not fixed: https://s3.amazonaws.com/clickhouse-test-reports/44813/f4dc6589011057316c1ed6d2819d29a0aec432b0/integration_tests__release__[2/4].html

azat commented:
>@azat The issues were not fixed: [https://s3.amazonaws.com/clickhouse-test-reports/44813/f4dc6589011057316c1ed6d2819d29a0aec432b0/integration_tests__release__[2/4].html](https://s3.amazonaws.com/clickhouse-test-reports/44813/f4dc6589011057316c1ed6d2819d29a0aec432b0/integration_tests__release__%5B2/4%5D.html)

@alexey-milovidov Don't see how this can be related

azat commented:
But likely you already fix the race in #44821
